File: base/modules/mcsrouter/mcsrouter/adapters/AgentAdapter.py

import os

import logging
logger = logging.getLogger(__name__)

import AdapterBase
import mcsrouter.utils.Timestamp

import mcsrouter.utils.TargetSystemManager
from mcsrouter import IPAddress
import mcsrouter.utils.PathManager as PathManager

def formatIPv6(i):
    assert ":" not in i
    assert len(i) == 32
    o = []
    countZero = 0
    while len(i) > 0:
        sub = i[:4]
        i = i[4:]

        while len(sub) > 0 and sub[0] == "0":
            sub=sub[1:]
        if sub == "":
            countZero += 1
            continue

        if countZero == 1:
            o.append("0")
        if countZero > 1:
            o.append("")

        countZero = 0
        o.append(sub)

    if countZero == 1:
        o.append("0")
    if countZero > 1:
        o.append(":")

    return ":".join(o)

class ComputerCommonStatus(object):
    """
    Class to represent the details that constitute the common computer status XML
    """
    def __init__(self, targetSystem):
        self.computerName = targetSystem.hostname()
        self.os = targetSystem.platform
        self.fqdn = IPAddress.getFQDN()
        self.user = "root@%s"% self.fqdn
        self.sls = targetSystem.checkIfUpdatableToSLS()
        self.arch = targetSystem.arch
        self.ipv4s = list(IPAddress.getNonLocalIPv4())
        self.ipv6s = list(IPAddress.getNonLocalIPv6())
        self.ipv6s = [ formatIPv6(i) for i in self.ipv6s ]

    def __eq__(self, other):
        return type(self) == type(other) and self.__dict__ == other.__dict__

    def __ne__(self, other):
        return not self == other

    def toStatusXml(self):
        ipv4 = self.ipv4s[0] if self.ipv4s else ""
        ipv6 = self.ipv6s[0] if self.ipv6s else ""

        result = [
            "<commonComputerStatus>",
            "<domainName>UNKNOWN</domainName>",
            "<computerName>%s</computerName>"%(self.computerName),
            "<computerDescription></computerDescription>",
            "<operatingSystem>%s</operatingSystem>"%self.os,
            "<lastLoggedOnUser>%s</lastLoggedOnUser>"%(self.user),
            "<ipv4>%s</ipv4>"%ipv4,
            "<ipv6>%s</ipv6>"%ipv6,
            "<fqdn>%s</fqdn>"%self.fqdn,
            "<processorArchitecture>%s</processorArchitecture>"%(self.arch)
            ]
        if self.ipv4s or self.ipv6s:
            result.append("<ipAddresses>")
            for ip in self.ipv4s:
                result.append("<ipv4>%s</ipv4>"%ip)
            for ip in self.ipv6s:
                result.append("<ipv6>%s</ipv6>"%ip)
            result.append("</ipAddresses>")

        result.append("</commonComputerStatus>")
        return "".join(result)


class AgentAdapter(AdapterBase.AdapterBase):
    def __init__(self, installdir=None):
        self.__m_last_status_time = None
        if installdir is not None:
            PathManager.INST = installdir

        self.__m_commonStatus = None

    def getAppId(self):
        return "AGENT"

    def getStatusTTL(self):
        return "PT10000S"

    def getTimestamp(self):
        return mcsrouter.utils.Timestamp.timestamp()

    def getStatusXml(self):
        return "".join((
            self.getStatusHeader(),
            self.getCommonStatusXml(),
            self.getPlatformStatus(),
            self.getStatusFooter()
            ))

    def getStatusHeader(self):
        return """<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<ns:computerStatus xmlns:ns="http://www.sophos.com/xml/mcs/computerstatus">
<meta protocolVersion="1.0" timestamp="%s"/>"""%(self.getTimestamp())

    def getStatusFooter(self):
        return """</ns:computerStatus>"""

    def __targetSystem(self):
        return mcsrouter.utils.TargetSystemManager.getTargetSystem(PathManager.installDir())

    def __createCommonStatus(self):
        ts = self.__targetSystem()
        assert ts is not None
        return ComputerCommonStatus(ts)

    def hasNewStatus(self):
        return self.__m_commonStatus != self.__createCommonStatus()

    def getCommonStatusXml(self):
        commonStatus = self.__createCommonStatus()

        if self.__m_commonStatus != commonStatus:
            self.__m_commonStatus = commonStatus

            logger.info("Reporting computerName=%s,fqdn=%s,IPv4=%s",
                self.__m_commonStatus.computerName,
                self.__m_commonStatus.fqdn,
                str(self.__m_commonStatus.ipv4s))

        return self.__m_commonStatus.toStatusXml()

    def getPlatformStatus(self):
        ts = self.__targetSystem()
        platform = ts.platform
        vendor = ts.vendor()
        kernel = ts.kernel
        osName = ts.osName
        return "".join((
            "<posixPlatformDetails>",
            "<platform>%s</platform>"%platform,
            "<vendor>%s</vendor>"%vendor,
            "<isServer>1</isServer>",
            "<osName>%s</osName>"%osName,
            "<kernelVersion>%s</kernelVersion>"%kernel,
            "</posixPlatformDetails>"))

File: base/modules/cmcsrouter/AgentAdapter.cpp
// Copyright 2022-2023 Sophos Limited. All rights reserved.

#include "AgentAdapter.h"

#include "Common/CurlWrapper/CurlWrapper.h"
#include "Common/HttpRequestsImpl/HttpRequesterImpl.h"
#include "Common/OSUtilities/IIPUtils.h"
#include "Common/OSUtilitiesImpl/LocalIPImpl.h"
#include "Common/OSUtilitiesImpl/PlatformUtils.h"
#include "Common/OSUtilitiesImpl/SystemUtils.h"
#include "Common/UtilityImpl/StringUtils.h"
#include "Common/UtilityImpl/TimeUtils.h"
#include "Common/XmlUtilities/Validation.h"

#include <iostream>

#define MAX_OS_NAME_LENGTH 255

namespace MCS
{
    AgentAdapter::AgentAdapter() :
        m_platformUtils(std::make_shared<Common::OSUtilitiesImpl::PlatformUtils>()),
        m_localIp(std::make_shared<Common::OSUtilitiesImpl::LocalIPImpl>())
    {
    }

    AgentAdapter::AgentAdapter(
        std::shared_ptr<Common::OSUtilities::IPlatformUtils> platformUtils,
        std::shared_ptr<Common::OSUtilities::ILocalIP> localIp) :
        m_platformUtils(std::move(platformUtils)), m_localIp(std::move(localIp))
    {
    }

    std::string AgentAdapter::getStatusXml(
        std::map<std::string, std::string>& configOptions,
        const std::shared_ptr<Common::HttpRequests::IHttpRequester>& client) const
    {
        std::stringstream statusXml;
        statusXml << getStatusHeader(configOptions) << getCommonStatusXml(configOptions)
                  << getCloudPlatformsStatus(client) << getPlatformStatus() << getStatusFooter();
        return statusXml.str();
    }

    std::string AgentAdapter::getStatusXml(std::map<std::string, std::string>& configOptions) const
    {
        std::shared_ptr<Common::CurlWrapper::ICurlWrapper> curlWrapper =
            std::make_shared<Common::CurlWrapper::CurlWrapper>();
        std::shared_ptr<Common::HttpRequests::IHttpRequester> client =
            std::make_shared<Common::HttpRequestsImpl::HttpRequesterImpl>(curlWrapper);
        return getStatusXml(configOptions, client);
    }

    std::string AgentAdapter::getStatusHeader(const std::map<std::string, std::string>& configOptions) const
    {
        std::stringstream header;
        header << R"(<?xml version="1.0" encoding="UTF-8" standalone="yes"?>)"
               << "<ns:computerStatus xmlns:ns=\"http://www.sophos.com/xml/mcs/computerstatus\">"
               << R"(<meta protocolVersion="1.0" timestamp=")"
               << Common::UtilityImpl::TimeUtils::MessageTimeStamp(std::chrono::system_clock::now())
               << "\" softwareVersion=\"";
        if (configOptions.find(MCS::VERSION_NUMBER) != configOptions.end())
        {
            header << configOptions.at(MCS::VERSION_NUMBER);
        }
        header << "\" />";
        return header.str();
    }

    std::string AgentAdapter::getCommonStatusXml(const std::map<std::string, std::string>& configOptions) const
    {
        std::stringstream commonStatusXml;

        std::vector<Common::OSUtilities::Interface> interfaces = m_localIp->getLocalInterfaces();
        m_platformUtils->sortInterfaces(interfaces);
        std::vector<std::string> ip4Addresses = m_platformUtils->getIp4Addresses(interfaces);
        std::vector<std::string> ip6Addresses = m_platformUtils->getIp6Addresses(interfaces);

        commonStatusXml << "<commonComputerStatus>"
                        << "<domainName>" << m_platformUtils->getDomainname() << "</domainName>"
                        << "<computerName>" << m_platformUtils->getHostname() << "</computerName>"
                        << "<computerDescription></computerDescription>"
                        << "<isServer>true</isServer>"
                        << "<operatingSystem>" << m_platformUtils->getPlatform() << "</operatingSystem>"
                        << "<lastLoggedOnUser>"
                        << "root@" << m_platformUtils->getHostname() << "</lastLoggedOnUser>"
                        << "<ipv4>" << m_platformUtils->getFirstIpAddress(ip4Addresses) << "</ipv4>"
                        << "<ipv6>" << m_platformUtils->getFirstIpAddress(ip6Addresses) << "</ipv6>"
                        << "<fqdn>" << m_platformUtils->getFQDN() << "</fqdn>"
                        << "<processorArchitecture>" << m_platformUtils->getArchitecture() << "</processorArchitecture>"
                        << getOptionalStatusValues(configOptions, ip4Addresses, ip6Addresses)
                        << "</commonComputerStatus>";
        return commonStatusXml.str();
    }

    std::string AgentAdapter::getOptionalStatusValues(
        const std::map<std::string, std::string>& configOptions,
        const std::vector<std::string>& ip4Addresses,
        const std::vector<std::string>& ip6Addresses) const
    {
        // For Groups, Products, and IP addresses
        std::stringstream optionals;

        if (configOptions.find(MCS::MCS_PRODUCTS) != configOptions.end())
        {
            std::string productsAsString =
                Common::UtilityImpl::StringUtils::replaceAll(configOptions.at(MCS::MCS_PRODUCTS), " ", "");
            optionals << "<productsToInstall>";
            if (productsAsString != "none")
            {
                std::vector<std::string> products =
                    Common::UtilityImpl::StringUtils::splitString(productsAsString, ",");
                for (const std::string& product : products)
                {
                    if (!product.empty() && Common::XmlUtilities::Validation::stringWillNotBreakXmlParsing(product))
                    {
                        optionals << "<product>" << product << "</product>";
                    }
                }
            }
            optionals << "</productsToInstall>";
        }

        if (configOptions.find(MCS::CENTRAL_GROUP) != configOptions.end())
        {
            std::string deviceGroupAsString = configOptions.at(MCS::CENTRAL_GROUP);
            if (Common::XmlUtilities::Validation::stringWillNotBreakXmlParsing(deviceGroupAsString))
            {
                optionals << "<deviceGroup>" << deviceGroupAsString << "</deviceGroup>";
            }
        }

        if (!ip4Addresses.empty() || !ip6Addresses.empty())
        {
            optionals << "<ipAddresses>";
            for (const std::string& ip4Address : ip4Addresses)
            {
                optionals << "<ipv4>" << ip4Address << "</ipv4>";
            }
            for (const std::string& ip6Address : ip6Addresses)
            {
                optionals << "<ipv6>" << ip6Address << "</ipv6>";
            }
            optionals << "</ipAddresses>";
        }

        std::vector<std::string> systemMacAddresses = m_platformUtils->getMacAddresses();
        if (!systemMacAddresses.empty())
        {
            optionals << "<macAddresses>";
            for (const std::string& macAddress : systemMacAddresses)
            {
                optionals << "<macAddress>" << macAddress << "</macAddress>";
            }
            optionals << "</macAddresses>";
        }
        return optionals.str();
    }

    std::string AgentAdapter::getCloudPlatformsStatus(
        const std::shared_ptr<Common::HttpRequests::IHttpRequester>& client) const
    {
        return m_platformUtils->getCloudPlatformMetadata(client);
    }

    std::string AgentAdapter::getPlatformStatus() const
    {
        std::stringstream platformStatusXml;
        std::string osName = m_platformUtils->getOsName();
        if (osName.size() > MAX_OS_NAME_LENGTH)
        {
            osName = osName.substr(0, MAX_OS_NAME_LENGTH);
        }
        platformStatusXml << "<posixPlatformDetails>"
                          << "<productType>sspl</productType>"
                          << "<platform>" << m_platformUtils->getPlatform() << "</platform>"
                          << "<vendor>" << m_platformUtils->getVendor() << "</vendor>"
                          << "<isServer>1</isServer>"
                          << "<osName>" << osName << "</osName>"
                          << "<kernelVersion>" << m_platformUtils->getKernelVersion() << "</kernelVersion>"
                          << "<osMajorVersion>" << m_platformUtils->getOsMajorVersion() << "</osMajorVersion>"
                          << "<osMinorVersion>" << m_platformUtils->getOsMinorVersion() << "</osMinorVersion>";

        std::shared_ptr<OSUtilities::ISystemUtils> systemUtils = std::make_shared<OSUtilities::SystemUtilsImpl>();
        if (systemUtils->getEnvironmentVariable("FORCE_UNINSTALL_SAV") == "1")
        {
            platformStatusXml << "<migratedFromSAV>1</migratedFromSAV>";
        }
        platformStatusXml << "</posixPlatformDetails>";
        return platformStatusXml.str();
    }

    std::string AgentAdapter::getStatusFooter() const
    {
        return "</ns:computerStatus>";
    }
} // namespace MCS
